src/server/gc_worker/compaction_filter.rs

```diff
@@ -199,138 +181,150 @@ impl CompactionFilterFactory for WriteCompactionFilterFactory {
-            deleting_filtered: None,
-            near_seek_distance: NEAR_SEEK_LIMIT,
-            write_iter: None,
-            write_iter_initialized: false,
-
             versions: 0,
             filtered: 0,
-            deleted: 0,
             total_versions: 0,
             total_filtered: 0,
-            total_deleted: 0,
+            mvcc_rollback_and_locks: 0,
+            orphan_versions: 0,
             versions_hist: MVCC_VERSIONS_HISTOGRAM.local(),
             filtered_hist: GC_DELETE_VERSIONS_HISTOGRAM.local(),
-            seek_times: 0,
-            next_times: 0,
-            mvcc_delete_skip_older: 0,
-            mvcc_rollback_and_locks: 0,
-            #[cfg(test)]
+            #[cfg(any(test, feature = "failpoints"))]
             callbacks_on_drop: {
                 let ctx = GC_CONTEXT.lock().unwrap();
                 ctx.as_ref().unwrap().callbacks_on_drop.clone()
             },
-        };
+        }
+    }
 
-        if filter.is_bottommost_level {
-            // MVCC-delete marks can only be handled at the bottommost level to forbid deleted
-            // RocksDB key/value pairs get exposed incorrectly. For example:
-            //   L0: |key_130_MVCC_DEL, key_110_MVCC_PUT(tombstone)|
-            //   L6: |key_110_MVCC_PUT|
-            // If key_130_MVCC_DEL is handled at L0 and the SST in L6 is not involved,
-            // key_110_MVCC_PUT could be exposed incorrectly.
-            let db = filter.engine.as_inner();
-            let cf = db.cf_handle(CF_WRITE).unwrap();
-            let iter = DBIterator::new_cf(db.clone(), cf, ReadOptions::new());
-            filter.write_iter = Some(RocksEngineIterator::from_raw(iter));
+    // `log_on_error` indicates whether to print an error log on scheduling failures.
+    // It's only enabled for `GcTask::OrphanVersions`.
+    fn schedule_gc_task(&self, task: GcTask, log_on_error: bool) {
+        if let Err(e) = self.gc_scheduler.schedule(task) {
+            if log_on_error {
+                error!("compaction filter schedule {} fail", e.0);
+            }
+        }
+    }
+
+    fn handle_bottommost_delete(&mut self) {
+        // Valid MVCC records should begin with `DATA_PREFIX`.
+        debug_assert_eq!(self.mvcc_key_prefix[0], keys::DATA_PREFIX);
+        let key = Key::from_encoded_slice(&self.mvcc_key_prefix[1..]);
+        self.mvcc_deletions.push(key);
+    }
+
+    fn gc_mvcc_deletions(&mut self) {
+        if !self.mvcc_deletions.is_empty() {
+            let empty = Vec::with_capacity(DEFAULT_DELETE_BATCH_COUNT);
+            let keys = mem::replace(&mut self.mvcc_deletions, empty);
+            let safe_point = self.safe_point.into();
+            self.schedule_gc_task(GcTask::GcKeys { keys, safe_point }, false);
         }
-        filter
     }
 
     fn do_filter(
         &mut self,
         _start_level: usize,
         key: &[u8],
-        sequence: u64,
+        _sequence: u64,
```

Comment:
seems it's useless now

Reply:
Yes, we can revert relative changes in `rust-rocksdb` later.